In [1]:
# Imports and pip installations (if needed)
import pandas

# Part 1: Load the dataset

In [2]:
# Load the given datasets

og_categorical_df = pandas.read_csv("chronic_kidney_disease_categorical.csv")
og_numerical_df = pandas.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print("CATEGORICAL DATA:")
print(og_categorical_df)
print("NUMERICAL DATA:")
print(og_numerical_df)

CATEGORICAL DATA:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  go

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The two csv files represent a list of patients, one on each row, with each box on that row corresponding to the information in its column. The ids for the patients in the two files sort of overlap, but I'm think that's a matter for a future question. One of the files contains the categorical information for each of the patients in it while the other contains the numerical information for each of the patients in it. Three examples of categorical columns are red blood cells, appetite, and bacteria, while three examples of numerical data are age, sodium, and potasium.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

def get_duplicate_count(df):
    col = "unique_id"
    return df[[col]].duplicated().sum()

print(get_duplicate_count(og_numerical_df))
# Drop these duplicate rows

numerical_df = og_numerical_df.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print(get_duplicate_count(og_categorical_df))
categorical_df = og_categorical_df.drop_duplicates()

print(categorical_df)


30
30
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
396     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
397     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
398     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
399     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
400     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad appet   pe  ane  Target  
0    yes  no  good  yes   no  

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: I don't think all the unique ids are present in each dataset because the original categorical tables is longer than the original numerical table and both have the same number of duplicates, meaning that even after removing duplicates, there are more unique ids in the categorical dataset than in the numerical dataset.

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df = categorical_df.merge(numerical_df, on="unique_id")

#Print the combined dataset
print(df)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
385     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
386     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
387     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
388     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
389     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad  ...    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value

percentage = df.isnull().any().sum()

# Print %
print(percentage)

# Drop these rows from the dataset

no_missing_df = df.dropna()

# Print the Dataset
print(no_missing_df)

23
     unique_id   al   su       rbc      pc         pcc          ba  htn   dm  \
4       881763  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
5       114717  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
7       901528  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
8       433613  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
9       621332  1.0  0.0  abnormal  normal  notpresent  notpresent   no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ...  ...   
382     996076  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
384     397388  2.0  2.0    normal  normal  notpresent     present  yes   no   
385     789906  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
386     490253  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
388     273038  0.0  0.0    normal  normal  notpresent  notpresent   no   no   

     cad  ...     bp    bgr    bu   

## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = no_missing_df.sort_values("Target")

# Print the dataset
print(sorted_df)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
118     534245  2.0  0.0    normal  abnormal     present  notpresent   no   
190     349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
337     343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
272     102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
329     514721  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
127     489160  0.0  0.0    normal    normal  notpresent  notpresent   no   
123     530138  0.0  0.0    normal    normal  notpresent  notpresent   no   
121     261137  0.0  0.0    normal    normal  notpresent  notpresent   no   
111     322882  0.0  0.0    normal    normal  notpresent  notpresent   no   
388     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_df = pandas.get_dummies(sorted_df, columns=categorical_columns, prefix=categorical_columns)

# Print the dataset
print(encoded_df)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
118     534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  3.8  ...   
190     349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
337     343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
272     102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
329     514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  4.8  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
127     489160  0.0  0.0  28.0  60.0   79.0   50.0   0.5  145.0  5.0  ...   
123     530138  0.0  0.0  23.0  80.0  111.0   34.0   1.1  145.0  4.0  ...   
121     261137  0.0  0.0  35.0  60.0  105.0   39.0   0.5  135.0  3.9  ...   
111     322882  0.0  0.0  33.0  80.0  128.0   38.0   0.6  135.0  3.9  ...   
388     273038  0.0  0.0  47.0  60.0  117.0   22.0   1.2  138.0  3.5  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

It would be the same to have a single column the each binary variable since one of the two columns as it currently is is redundant (simply being the opposite of its other column) so a single column for each binary variable would hold the same amount of information. I'm not sure by what you mean as the effect that "this" would have for a categorical variables of more than two values. If you mean changing it to a single column of booleans, then this would obviously remove a considerable amount of information; like if your condiments are "ketchup", "mustard", "relish", "mayonase", "pickles", then if you only keep the "ketchup" column, then all the infomation you're gonna have in terms of condiment is whether it's ketchup or non-ketchup. You could however remove only one column and infer that it is true if all the other columns are false and false otherwise.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [8]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# def get_func(col):
#     mean = col.mean()
#     sdev3 = col.std() * 3
#     bottom = mean - sdev3
#     top = mean + sdev3
#     print(bottom)
#     print(top)
#     return lambda n: n < bottom or n > top

# funcs = encoded_df.apply(get_func)

# def is_not_outlier(row):
#     for nc in numerical_columns:
#         if not funcs[nc]:
#             return False
#     return True

filtered_df = encoded_df.copy()

for nc in numerical_columns:
    mean = encoded_df[[nc]].mean()[nc]
    sdev3 = encoded_df[[nc]].std()[nc] * 3
    filtered_df = filtered_df[(filtered_df[nc] - mean).abs() <= sdev3]

#filtered_df = encoded_df[(encoded_df - mean).abs() <= 3 * sdev].dropna()

# Print the dataset
print(filtered_df)

     unique_id   al   su   age    bp    bgr     bu   sc    sod  pot  ...  \
337     343710  2.0  0.0  61.0  80.0  173.0  148.0  3.9  135.0  5.2  ...   
329     514721  2.0  0.0  56.0  90.0  129.0  107.0  6.7  131.0  4.8  ...   
163     124304  3.0  4.0  69.0  70.0  264.0   87.0  2.7  130.0  4.0  ...   
314     980291  3.0  0.0  62.0  70.0  122.0   42.0  1.7  136.0  4.7  ...   
277     615697  4.0  0.0   6.0  60.0   94.0   67.0  1.0  135.0  4.9  ...   
..         ...  ...  ...   ...   ...    ...    ...  ...    ...  ...  ...   
127     489160  0.0  0.0  28.0  60.0   79.0   50.0  0.5  145.0  5.0  ...   
123     530138  0.0  0.0  23.0  80.0  111.0   34.0  1.1  145.0  4.0  ...   
121     261137  0.0  0.0  35.0  60.0  105.0   39.0  0.5  135.0  3.9  ...   
111     322882  0.0  0.0  33.0  80.0  128.0   38.0  0.6  135.0  3.9  ...   
388     273038  0.0  0.0  47.0  60.0  117.0   22.0  1.2  138.0  3.5  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
337   F

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [9]:
# Normalize the all Numerical Attributes in the dataset.

normalized_df = filtered_df.copy()
normalized_df[numerical_columns] = (filtered_df[numerical_columns] - filtered_df[numerical_columns].mean()) / filtered_df[numerical_columns].std()


# Print the dataset
print(normalized_df)

     unique_id   al   su       age        bp       bgr        bu        sc  \
337     343710  2.0  0.0  0.796206  0.723662  1.137875  3.951101  1.655981   
329     514721  2.0  0.0  0.482592  1.744405  0.163875  2.431617  3.520699   
163     124304  3.0  4.0  1.297989 -0.297082  3.152284  1.690405  0.856816   
314     980291  3.0  0.0  0.858929 -0.297082  0.008921  0.022679  0.190845   
277     615697  4.0  0.0 -2.653552 -1.317826 -0.610898  0.949194 -0.275334   
..         ...  ...  ...       ...       ...       ...       ...       ...   
127     489160  0.0  0.0 -1.273649 -1.317826 -0.942943  0.319164 -0.608319   
123     530138  0.0  0.0 -1.587263  0.723662 -0.234579 -0.273806 -0.208737   
121     261137  0.0  0.0 -0.834589 -1.317826 -0.367398 -0.088503 -0.608319   
111     322882  0.0  0.0 -0.960035  0.723662  0.141739 -0.125563 -0.541722   
388     273038  0.0  0.0 -0.081914 -1.317826 -0.101761 -0.718533 -0.142140   

          sod       pot  ...  cad_no  cad_yes  appet_good  appe

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: None of of the columns which are simply the opposite of another column (such as "cad_no" vs "cad_yes" or "pc_abnormal" vs "pc_normal") are necessary because their information can be easily deduced by their corresponding column.

In [10]:
#Remove that column
final_df = normalized_df.drop(columns = ["rbc_abnormal", "pc_abnormal", "pcc_notpresent", "ba_notpresent", "htn_no", "dm_no", "cad_no", "appet_poor", "pe_no", "ane_no", "Target_notckd"])

# Print the dataset
print(final_df)

     unique_id   al   su       age        bp       bgr        bu        sc  \
337     343710  2.0  0.0  0.796206  0.723662  1.137875  3.951101  1.655981   
329     514721  2.0  0.0  0.482592  1.744405  0.163875  2.431617  3.520699   
163     124304  3.0  4.0  1.297989 -0.297082  3.152284  1.690405  0.856816   
314     980291  3.0  0.0  0.858929 -0.297082  0.008921  0.022679  0.190845   
277     615697  4.0  0.0 -2.653552 -1.317826 -0.610898  0.949194 -0.275334   
..         ...  ...  ...       ...       ...       ...       ...       ...   
127     489160  0.0  0.0 -1.273649 -1.317826 -0.942943  0.319164 -0.608319   
123     530138  0.0  0.0 -1.587263  0.723662 -0.234579 -0.273806 -0.208737   
121     261137  0.0  0.0 -0.834589 -1.317826 -0.367398 -0.088503 -0.608319   
111     322882  0.0  0.0 -0.960035  0.723662  0.141739 -0.125563 -0.541722   
388     273038  0.0  0.0 -0.081914 -1.317826 -0.101761 -0.718533 -0.142140   

          sod       pot  ...  pc_normal  pcc_present  ba_presen

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file
normalized_df.to_csv("kidneys.csv")

# Export the dataframe to a new json file
normalized_df.to_json("kidneys.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

| unique_id |    al |   su |                  age |         bp |         bgr |         bu |         sc |         sod |         pot |         hemo |         pcv |         wbcc |         rbcc |          rbc_abnormal |          rbc_normal |          pc_abnormal |          pc_normal |          pcc_notpresent |          pcc_present |            ba_notpresent |            ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----------|-------|-------|---------------------|------------|--------------|------------|------------|-------------|-------------|---------------|--------------|---------------|---------------|------------------------|---------------------|-----------------------|----------------------|----------------------|-------------------|-------------------------|------------------|--------|---------|-------|--------|--------|---------|------------|-------------|-------|--------|--------|---------|------------|-----------------|
|       337 | 343710 |  2.0 |                 0.0 | 0.796206132 | 0.723661532 | 1.137875238 | 3.951101019 | 1.655980773 | -0.934242441 | 1.298120661  | -2.738870477 | -2.712933250 | 0.506560146 |                  True |                False |                  True |                False |                  True |              False |                   True |               False |  False |   True |  False |   True |  False |   True |        True |        False |  False |   True |  False |   True |       True |         False |
|       329 | 514721 |  2.0 |                 0.0 | 0.482591724 | 1.744405167 | 0.163875179 | 2.431616979 | 3.520698617 | -1.629734036 | 0.692029481  | -2.160513767 | -2.030370247 | -0.726003841 |                  True |                False |                  True |                False |                  True |              False |                   True |               False |  False |   True |  True |   False |  True |   False |        True |        False |  True |   False |  True |   False |       True |         False |
|       163 | 124304 |  3.0 |                 4.0 | 1.297989186 | -0.297082103 | 3.152284450 | 1.690405252 | 0.856815982 | -1.803606935 | -0.520152878 | -0.755933184 | -0.938269442 | 0.682640716 |                 False |                 True |                  True |                False |                  True |              False |                   True |               False |  False |   True |  False |   True |  False |   True |        True |        False |  False |   True |  False |   True |       True |         False |
|       314 | 980291 |  3.0 |                 0.0 | 0.858929014 | -0.297082103 | 0.008920625 | 0.022678866 | 0.190845323 | -0.760369542 | 0.540506686  | -0.714621991 | -0.665244241 | -0.065701705 |                 False |                 True |                  True |                False |                  True |              False |                   True |               False |  False |   True |  False |   True |  True |   False |        True |        False |  True |   False |  True |   False |       True |         False |
|       277 | 615697 |  4.0 |                 0.0 | -2.653552360 | -1.317825737 | -0.610897594 | 0.949193525 | -0.275334138 | -0.934242441 | 0.843552276  | -1.830024218 | -1.893857646 | 3.808070827 |                  True |                False |                  True |                False |                  True |              False |                   False |                True |  True |   False |  True |   False |  True |   False |       False |         True |  True |   False |  True |   False |       True |         False |
|       312 | 955830 |  3.0 |                 1.0 | 0.419868842 | 0.723661532 | 2.045466201 | 1.171557043 | 1.655980773 | -0.586496644 | 0.843552276  | -1.416912282 | -1.347807244 | -0.285802417 |                 False |                 True |                  True |                False |                 False |               True |                   False |                True |  False |   True |  False |   True |  False |   True |        True |        False |  False |   True |  False |   True |       True |         False |
|       307 | 546225 |  1.0 |                 0.0 | -0.144637093 | -1.317825737 | 0.916511588 | 1.875708184 | 1.256398377 | 0.108994951  | -0.520152878 | -1.871335412 | -2.166882848 | 2.883647836 |                 False |                 True |                  False |                True |                  True |              False |                   True |               False |  False |   True |  False |   True |  True |   False |        True |        False |  True |   False |  True |   False |       True |         False |
|       122 | 475200 |  4.0 |                 2.0 | 0.984374777 | 2.765148802 | 0.916511588 | 0.467405902 | 3.853683947 | -0.064877947 | 0.388983891  | -2.656248090 | -2.439908049 | -0.241782275 |                  True |                False |                  True |                False |                  True |              False |                   False |                True |  False |   True |  False |   True |  True |   False |        True |        False |  True |   False |  True |   False |       True |         False |
|       159 | 234299 |  4.0 |                 0.0 | -0.019191329 | 0.723661532 | -0.588761229 | 4.507009814 | 4.186669276 | -0.760369542 | -0.823198468 | -1.871335412 | -1.620832445 | -0.505903129 |                 False |                 True |                  True |                False |                  True |              False |                   False |                True |  True |   False |  True |   False |  True |   False |       True |        False |  True |   False |  True |   False |       True |         False |
|       124 | 995177 |  4.0 |                 3.0 | 1.297989186 | -0.297082103 | 2.045466201 | 2.023950529 | 3.254310354 | -3.542335923 | -0.671675673 | -2.036580186 | -2.166882848 | 1.519023422 |                 False |                 True |                  True |                False |                 False |               True |                   False |                True |  False |   True |  False |   True |  False |   True |        True |        False |  False |   True |  True |   False |       True |         False |
|       303 | 137148 |  3.0 |                 2.0 | 1.548880712 | 2.765148802 | 3.838511763 | 1.801587011 | 2.788130893 | -0.064877947 | -2.186903622 | -2.119202573 | -1.893857646 | -0.461882987 |                  True |                False |                  True |                False |                 False |               True |                   False |                True |  False |   True |  False |   True |  False |   True |        True |        False |  True |   False |  True |   False |       True |         False |
|       293 | 474407 |  4.0 |                 0.0 | -1.712709135 | 1.744405167 | -0.323124850 | -0.051442306 | 0.190845323 | -2.672971429 | -1.277766853 | -2.491003316 | -2.849445851 | 1.915204703 |                 False |                 True |                  True |                False |                 False |               True |                   False |                True |  True |   False |  True |   False |  True |   False |        True |        False |  True |   False |  True |   False |       True |         False |
|       290 | 923613 |  4.0 |                 1.0 | 0.984374777 | -1.317825737 | 2.598875325 | 0.615648248 | 1.922369036 | -0.586496644 | 1.601166251  | -1.995268992 | -2.030370247 | -0.241782275 |                  True |                False |                  True |                False |                  True |              False |                   False |                True |  False |   True |  False |   True |  True |   False |        True |        False |  True |   False |  True |   False |       True |         False |
|       264 | 197452 |  4.0 |                 1.0 | 0.482591724 | -0.297082103 | 1.956920742 | -0.570290515 | 0.190845323 | -0.760369542 | -0.823198468 | 0.731269785  | 1.109419568  | 1.959224846 |                  True |                False |                  False |                True |                  True |              False |                   True |               False |  True |   False |  True |   False |  True |   False |        True |        False |  True |   False |  True |   False |       True |         False |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

You could ask the language model to write code that converts the data into a table instead and then run that code on the large dataset.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [12]:
# Code to rename all the columns in the dataset
from bs4 import BeautifulSoup

html_table = "<tbody><tr><td>age</td><td>Feature</td><td>Integer</td><td>Age</td><td></td><td>year</td><td>yes</td> </tr><tr><td>bp</td><td>Feature</td><td>Integer</td><td></td><td>blood pressure</td><td>mm/Hg</td><td>yes</td> </tr><tr><td>sg</td><td>Feature</td><td>Categorical</td><td></td><td>specific gravity</td><td></td><td>yes</td> </tr><tr><td>al</td><td>Feature</td><td>Categorical</td><td></td><td>albumin</td><td></td><td>yes</td> </tr><tr><td>su</td><td>Feature</td><td>Categorical</td><td></td><td>sugar</td><td></td><td>yes</td> </tr><tr><td>rbc</td><td>Feature</td><td>Binary</td><td></td><td>red blood cells</td><td></td><td>yes</td> </tr><tr><td>pc</td><td>Feature</td><td>Binary</td><td></td><td>pus cell</td><td></td><td>yes</td> </tr><tr><td>pcc</td><td>Feature</td><td>Binary</td><td></td><td>pus cell clumps</td><td></td><td>yes</td> </tr><tr><td>ba</td><td>Feature</td><td>Binary</td><td></td><td>bacteria</td><td></td><td>yes</td> </tr><tr><td>bgr</td><td>Feature</td><td>Integer</td><td></td><td>blood glucose random</td><td>mgs/dl</td><td>yes</td> </tr><tr><td>bu</td><td>Feature</td><td>Integer</td><td></td><td>blood urea</td><td>mgs/dl</td><td>yes</td> </tr><tr><td>sc</td><td>Feature</td><td>Continuous</td><td></td><td>serum creatinine</td><td>mgs/dl</td><td>yes</td> </tr><tr><td>sod</td><td>Feature</td><td>Integer</td><td></td><td>sodium</td><td>mEq/L</td><td>yes</td> </tr><tr><td>pot</td><td>Feature</td><td>Continuous</td><td></td><td>potassium</td><td>mEq/L</td><td>yes</td> </tr><tr><td>hemo</td><td>Feature</td><td>Continuous</td><td></td><td>hemoglobin</td><td>gms</td><td>yes</td> </tr><tr><td>pcv</td><td>Feature</td><td>Integer</td><td></td><td>packed cell volume</td><td></td><td>yes</td> </tr><tr><td>wbcc</td><td>Feature</td><td>Integer</td><td></td><td>white blood cell count</td><td>cells/cmm</td><td>yes</td> </tr><tr><td>rbcc</td><td>Feature</td><td>Continuous</td><td></td><td>red blood cell count</td><td>millions/cmm</td><td>yes</td> </tr><tr><td>htn</td><td>Feature</td><td>Binary</td><td></td><td>hypertension</td><td></td><td>yes</td> </tr><tr><td>dm</td><td>Feature</td><td>Binary</td><td></td><td>diabetes mellitus</td><td></td><td>yes</td> </tr><tr><td>cad</td><td>Feature</td><td>Binary</td><td></td><td>coronary artery disease</td><td></td><td>yes</td> </tr><tr><td>appet</td><td>Feature</td><td>Binary</td><td></td><td>appetite</td><td></td><td>yes</td> </tr><tr><td>pe</td><td>Feature</td><td>Binary</td><td></td><td>pedal edema</td><td></td><td>yes</td> </tr><tr><td>ane</td><td>Feature</td><td>Binary</td><td></td><td>anemia</td><td></td><td>yes</td> </tr><tr><td>class</td><td>Target</td><td>Binary</td><td></td><td>ckd or not ckd</td><td></td><td>no</td> </tr></tbody>"

# Parse the HTML
soup = BeautifulSoup(html_table, 'html.parser')

# Initialize an empty dictionary
col_names = {}

# Find all rows in the table
rows = soup.find_all('tr')

# Iterate through the rows
for row in rows:
    columns = row.find_all('td')
    # Check if we have enough columns
    if len(columns) >= 5:
        key = columns[0].text.strip()  # 1st column
        value = columns[4].text.strip()  # 5th column
        col_names[key] = value

col_names["age"] = "age"
col_names["Target"] = "Target"

for col_name in final_df.columns.values:
    if col_name != "unique_id" and "_" in col_name:
        pair = col_name.split("_")
        col_names[col_name] = col_names[pair[0]] + " " + pair[1]

renamed_df = final_df.rename(columns = col_names)

print(renamed_df)

     unique_id  albumin  sugar       age  blood pressure  \
337     343710      2.0    0.0  0.796206        0.723662   
329     514721      2.0    0.0  0.482592        1.744405   
163     124304      3.0    4.0  1.297989       -0.297082   
314     980291      3.0    0.0  0.858929       -0.297082   
277     615697      4.0    0.0 -2.653552       -1.317826   
..         ...      ...    ...       ...             ...   
127     489160      0.0    0.0 -1.273649       -1.317826   
123     530138      0.0    0.0 -1.587263        0.723662   
121     261137      0.0    0.0 -0.834589       -1.317826   
111     322882      0.0    0.0 -0.960035        0.723662   
388     273038      0.0    0.0 -0.081914       -1.317826   

     blood glucose random  blood urea  serum creatinine    sodium  potassium  \
337              1.137875    3.951101          1.655981 -0.934242   1.298121   
329              0.163875    2.431617          3.520699 -1.629734   0.692029   
163              3.152284    1.690405  

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [13]:
# Converted SQL to Pandas code

#I'm gonna change the name cause there's no column named "Target" anymore after we did that encoding stuff
count_df = renamed_df.groupby('Target ckd').size().reset_index(name='count')
print(count_df)

   Target ckd  count
0       False    111
1        True     23
